service/src/main/kotlin/fi/espoo/evaka/reports/MealReportController.kt

```diff
@@ -0,0 +1,232 @@
+    val presentTimeRanges =
+        if (fixedScheduleRange != null) listOf(fixedScheduleRange)
+        else reservations.filterIsInstance<ReservationResponse.Times>().map { it.range }
+    // if we don't have data about when child will be present, default to breakfast + lunch + snack
+    if (presentTimeRanges.isEmpty()) {
+        return sequenceOf(
+            MealInfo(MealType.BREAKFAST),
+            MealInfo(MealType.LUNCH),
+            MealInfo(MealType.SNACK)
+        )
+    }
+    // otherwise check unit meal times against the present time ranges
+    val meals = mutableListOf<MealInfo>()
+
+    fun addMealIfPresent(mealTime: TimeRange?, mealType: MealType) {
+        if (mealTime != null && presentTimeRanges.any { it.overlaps(mealTime) }) {
+            meals.add(MealInfo(mealType))
+        }
+    }
+
+    addMealIfPresent(daycare.mealtimeBreakfast, MealType.BREAKFAST)
+    addMealIfPresent(
+        daycare.mealtimeLunch,
+        if (usePreschoolMealTypes) MealType.LUNCH_PRESCHOOL else MealType.LUNCH
+    )
+    addMealIfPresent(daycare.mealtimeSnack, MealType.SNACK)
+    addMealIfPresent(daycare.mealtimeSupper, MealType.SUPPER)
+    addMealIfPresent(daycare.mealtimeEveningSnack, MealType.EVENING_SNACK)
+
+    return meals.asSequence()
+}
+
+private fun getMealReport(db: Database.Read, date: LocalDate, unitId: DaycareId): MealReportData {
+    val daycare = db.getDaycare(unitId) ?: throw BadRequest("Daycare not found for $unitId")
+    if (!daycare.operationDays.contains(date.dayOfWeek.value))
+        return MealReportData(date, daycare.name, emptyList())
+
+    val preschoolTerms = db.getPreschoolTerms()
+
+    val childrenToPlacementTypeMap = db.childPlacementsForDay(unitId, date)
+    val childrenReservationsAndAttendances =
+        db.getChildData(unitId, childrenToPlacementTypeMap.keys, date.toFiniteDateRange())
+
+    val mealInfoMap =
+        childrenToPlacementTypeMap
+            .asSequence()
+            .flatMap { (childId, placementType) ->
+                val childReservationsAndAttendances = childrenReservationsAndAttendances[childId]!!
+                val fixedScheduleRange =
+                    placementType.fixedScheduleOnlyRange(
+                        date,
+                        daycare.dailyPreschoolTime,
+                        daycare.dailyPreparatoryTime,
+                        preschoolTerms
+                    )
+                val absent =
+                    childReservationsAndAttendances.absences[date]?.size ==
+                        placementType.absenceCategories().size
+                val usePreschoolMealTypes = preschoolPlacementTypes.contains(placementType)
+                childMeals(
+                    fixedScheduleRange,
+                    childReservationsAndAttendances.reservations[date] ?: emptyList(),
+                    absent,
+                    daycare,
+                    usePreschoolMealTypes
+                )
+            }
+            .groupBy { it }
```

Comment:
Tämä selventäisi ryhmittelyn tarkoitusta:

```suggestion
            .groupBy { it.mealType }
```

Reply:
Tällä hetkellä se olisi OK, mutta seuraavassa vaiheessa `MealInfo`on populoidaan myös erikoisruokavalio ja lapsen nimitieto. Silloin kunkin eroavan `MealInfo`n on oltava raportissa / tilauksessa omalla rivillään.